main.c

```diff
@@ -0,0 +1,91 @@
+#include <string.h>
+#include "main.h"
+#include "_cgo_export.h"
+
+result *save_to_png(size_t max_store, int page_number, int width, float scale, const unsigned char *payload, size_t payload_length) {
+	if (max_store == 0) {
+		max_store = FZ_STORE_DEFAULT;
+	}
+
+	fz_context *ctx = fz_new_context(NULL, NULL, max_store);
+	if (!ctx) {
+		errorHandler("fail to create mupdf context");
+		return NULL;
+	}
+
+	result *r = malloc(sizeof(result));
+	r->context = ctx;
+	r->buffer = NULL;
+
+	int exit = 0;
+	fz_stream *stream = NULL;
+	fz_document *doc = NULL;
+	fz_page *page = NULL;
+	fz_display_list *list = NULL;
+	fz_device *device = NULL;
+	fz_pixmap *pixmap = NULL;
+	fz_device *draw_device = NULL;
+	fz_buffer *buffer = NULL;
+	fz_try(ctx) {
+		fz_register_document_handlers(ctx);
+		stream = fz_open_memory(ctx, payload, payload_length);
+		doc = fz_open_document_with_stream(ctx, "document.pdf", stream);
+		page = fz_load_page(ctx, doc, page_number);
+
+		float scale_factor = 1;
+		fz_rect bounds = fz_bound_page(ctx, page);
+		if (width != 0) {
+			scale_factor = width / bounds.x1;
+		} else if (scale != 0) {
+			scale_factor = scale;
+		}
+
+		fz_matrix ctm = fz_scale(scale_factor, scale_factor);
+		bounds = fz_transform_rect(bounds, ctm);
+		fz_irect bbox = fz_round_rect(bounds);
+		list = fz_new_display_list(ctx, bounds);
+		device = fz_new_list_device(ctx, list);
+		fz_run_page(ctx, page, device, fz_identity, NULL);
+		pixmap = fz_new_pixmap_with_bbox(ctx, fz_device_rgb(ctx), bbox, NULL, 1);
+		fz_clear_pixmap_with_value(ctx, pixmap, 0xff);
+		draw_device = fz_new_draw_device(ctx, ctm, pixmap);
+		fz_run_display_list(ctx, list, draw_device, fz_identity, bounds, NULL);
+		buffer = fz_new_buffer_from_pixmap_as_png(ctx, pixmap, fz_default_color_params);
+		r->buffer = buffer;
+
+		size_t len = fz_buffer_storage(ctx, buffer, NULL);
+		const char *output = fz_string_from_buffer(ctx, buffer);
+		r->len = len;
+		r->data = output;
```

Comment:
Why are you returning both r->buffer (which is the png data) and r->data. It seems to me you would only need the png data, no?

Reply:
The documentation says that if the buffer is freed it will also free the underlying storage. That's why I'm keeping the buffer to free it after it's used at the Go layer.